---
title: "hw16"
author: "Meha Thakur (A16020450)"
format: pdf
toc: true
---

```{r}
library(BiocManager)
#BiocManager::install("tximport")
library(tximport)
```

## Downstream analysis

since each kallisto has its own output, need to import using tximport

```{r}
# setup the folder and filenames to read
folders <- dir(pattern="SRR21568*")
samples <- sub("_quant", "", folders)
files <- file.path( folders, "abundance.h5" )
names(files) <- samples

txi.kallisto <- tximport(files, type = "kallisto", txOut = TRUE)

head(txi.kallisto$counts) 
```
txi.kallisto imports the each gene (rows) and each dataset(columns?)

how many counts for each sample (how many transcripts). We can also detect how many transcripts are detected in at least one sample
```{r}
colSums(txi.kallisto$counts)

sum(rowSums(txi.kallisto$counts)>0)
```

now we filter out annotated samples with no reads, and no change
```{r}
#no reads
to.keep <- rowSums(txi.kallisto$counts) > 0
kset.nonzero <- txi.kallisto$counts[to.keep,]

#no change
keep2 <- apply(kset.nonzero,1,sd)>0 #row-wise standard deviations only keeps rows with nonzero variability
x <- kset.nonzero[keep2,]
```


```{r}

```


## PCA analysis

```{r}
pca <- prcomp(t(x), scale=TRUE)
summary(pca)
```
use the first two PCs to visualse transcriptomic profiles
```{r}

plot(pca$x[,1], pca$x[,2],
     col=c("blue","blue","red","red"),
     xlab="PC1", ylab="PC2", pch=16)


```

using ggplot for PC1 vs PC2, PC1 vs. PC3
```{r}
library(ggplot2)
library(ggrepel)

# Make metadata object for the samples
colData <- data.frame(condition = factor(rep(c("control", "treatment"), each = 2)))
rownames(colData) <- colnames(txi.kallisto$counts)

# Make the data.frame for ggplot 
y <- as.data.frame(pca$x)
y$Condition <- as.factor(colData$condition)

ggplot(y) +
  aes(PC1, PC2, col=Condition) +
  geom_point() +
  geom_text_repel(label=rownames(y)) +
  theme_bw()
```

DESeq analysis

```{r}
library(DESeq2)


```
```{r}
sampleTable <- data.frame(condition = factor(rep(c("control", "treatment"), each = 2)))
rownames(sampleTable) <- colnames(txi.kallisto$counts)
```


```{r}
dds <- DESeqDataSetFromTximport(txi.kallisto,
                                sampleTable, 
                                ~condition)

```

DESeq
```{r}
dds <- DESeq(dds)
```
```{r}
res <- results(dds)
head(res)
```


